### PR TITLE
SentencePiece changes + UnitTests + Making tokenizer APIs consistent

### PR DESCRIPTION
**Pull request by aklife97**
### Summary of the PR ###
Broadly, this PR does the following:
* Changes `SentencePiece` wrapper and `create_spt_model` function to handle special tokens inside SentencePiece model (instead of inside the wrapper that happens currently).
* Makes all TokenizerSpec tokenizer wrappers (AutoTokenizer, CharTokenizer, WordTokenizer, YTTM, SentecePiece) APIs consistent - except for a few differences as discussed later.
* Adds tests for the current SentencePiece and YTTM tokenizers.

### SentencePiece Changes ###
Current implementation of the SentencePiece wrapper handles special tokens inside the wrapper using a special token dictionary - rather than inside the tokenizer model. This makes the special tokens created non-persistent and volatile to change during different runs. 

To better handle this situation, the `create_spt_model` function now supports the addition of `bos`, `eos`, and `pad` tokens directly and special tokens can only be added during train time. More custom special tokens can be passed as `Control symbols` or `User defined symbols` as defined by sentencepiece. `bos`, `eos`, and `pad` are by default `custom symbols`. The major difference between setting a special token as `control` vs `user defined` is that `user defined symbols` can encoded and decoded directly from the text, whereas `control symbols` can only be added manually.  For more information on the differnce, refer [SentencePiece](https://github.com/google/sentencepiece/blob/master/doc/special_symbols.md).  Despite these changes, any previously coded calls to `create_spt_model` would work in the **_same fashion and do not need to be changed_**. 

The SentencePiece wrapper has also been changed to accommodate the above. While the default behavior of the SentencePiece wrapper has been changed with this, older behavior for previously trained models can be brought back by setting `legacy = True` while instantiating the wrapper.

### Making TokenizerSpec tokenizer APIs consistent ###

This PR also tries to make all the tokenizer wrappers that inherit from TokenizerSpec consistent. It changes the YTTM tokenizer's `ids_to_tokens` function to make it consistent with others. Again, while this has been set as the default behavior, previous behavior can be restored by setting `legacy = True`.

While the tokenizers are now mostly similar, there are some exceptions:
* AutoTokenizer differs from all other tokenizers in the fact that the AutoTokenizer's `tokens_to_text` function _**does not**_ remove special tokens. This is probably a bug, and should ideally be addressed when it's certain that it won't break anything. 
* YTTM and SentecePiece tokenizers do not encode special tokens (control symbols for SentencePiece) when `text_to_ids` or `text_to_tokens` is called. While this should ideally be consistent across all tokenizers, as this inhibits user text to change the programmatic meaning of the text, other tokenizers (char, word, and auto) encode the special tokens. An example:
When trying to tokenize (`text_to_tokens`)
`text: a b [UNK] c` 
yttm and sentencepice: `['a', 'b', '[', 'U', 'N', 'K', ']', 'c']`
others: `['a', 'b', '[UNK]', 'c']`

#### Behavior of TokenizerSpec API ####
* `[tokens|ids]_to_[ids|tokens]`: Converts tokens to ids and vice versa while preserving all special token infomation, i.e. `len(ids) == len(tokens)`
* `text_to_[tokens|ids]`: Encodes text as tokens/ids. Exception: Auto/word/char encode special tokens in text as is.
* `[tokens|id]_to_text`:  Converts tokens/ids to text and removes special tokens (only control symbols for SentencePiece). Exception: `tokens_to_text` of `AutoTokenizer`.

### SentencePiece and YTTM tests ###
Added tests for testing YTTM and SentencePiece.

### CODE REVIEW

File: nemo/collections/common/tokenizers/sentencepiece_tokenizer.py
Comment (titu1994) [inline nemo/collections/common/tokenizers/sentencepiece_tokenizer.py:266]: Document difference between `control_symbols` and `user_defined_symbols` - maybe with example.

**Review by titu1994 — COMMENTED** [review]
SentencePiece changes look fine. @yzhang123  for final approval

File: nemo/collections/common/tokenizers/sentencepiece_tokenizer.py
Comment (aklife97) [reply nemo/collections/common/tokenizers/sentencepiece_tokenizer.py:266]: Added explanation for both in the docstring.

**Comment by ericharper** [comment]
Doesn't have to be this PR, but I think the contents of the README would be a great addition to the NeMo NLP Docs.

File: nemo/collections/common/tokenizers/sentencepiece_tokenizer.py
Comment (okuchaiev) [inline nemo/collections/common/tokenizers/sentencepiece_tokenizer.py:39]: can you please add "legacy" to docstring. Say something brief.

File: nemo/collections/common/tokenizers/sentencepiece_tokenizer.py
Comment (okuchaiev) [inline nemo/collections/common/tokenizers/sentencepiece_tokenizer.py:54]: let's remove the suggestion: ", or set legacy=True."

**Review by okuchaiev — APPROVED** [review]
looks good to me. few small comments

File: nemo/collections/common/tokenizers/sentencepiece_tokenizer.py
Comment (aklife97) [reply nemo/collections/common/tokenizers/sentencepiece_tokenizer.py:39]: Addressed both comments. Waiting for @yzhang123 's review. Will merge once approved.

File: nemo/collections/nlp/modules/common/tokenizer_utils.py
Comment (yzhang123) [inline nemo/collections/nlp/modules/common/tokenizer_utils.py:85]: could you add docstring for bpe_dropout?

**Review by yzhang123 — CHANGES_REQUESTED** [review]

File: nemo/collections/common/tokenizers/sentencepiece_tokenizer.py
Comment (yzhang123) [reply nemo/collections/common/tokenizers/sentencepiece_tokenizer.py:39]: could you move docstring outside of constructor for automatic doc generation?

Comment (aklife97) [reply nemo/collections/common/tokenizers/sentencepiece_tokenizer.py]: Moved the args docstring outside. Is it fine now?

File: nemo/collections/nlp/modules/common/tokenizer_utils.py
Comment (aklife97) [reply nemo/collections/nlp/modules/common/tokenizer_utils.py:88]: Added.

**Review by yzhang123 — APPROVED** [review]

**Review by yzhang123 — APPROVED** [review]

**Pull request merged**
